# Module 1: Technology
## Lecture 1: 

### Scope of course (Brazil)

### Three phase diagram

### open bioinformatics Spell

### Intro to Sequencing Technology

### Sequenceing data file formats: FASTA/FASTQ

### Unix Crash Course

## Lab 1:

### Unix Intro

### (Prog) Code a FASTA/FASTQ Parser

### (Bioinfo) Use FASTX toolkit to manipulate FASTA/FASTQ files

### (R) FASTQ statistics exploration

# Module 1: Alignment (mappers, file formats)
## Lecture 2: 

### Intro to alignment programs
* Indexing
	* Genome conventions

* Mapping
	* DNA
	* RNA

### SAM/BAM file format

## Lab 2:

### (Prog) Code a rudimentary mapping program

### (Bioinfo 1): DNA/ChIP-seq: DNA mappers

* Index and map DNA data
	 
### (Bioinfo 2): RNA/RNA-seq: Spliced mappers

* Index (?) and map RNAseq data
	
### (R) SAM file stats

# Module 3: Processing alignments

## Lecture 3:

* Samtools

* PICARD

* Other imported Bioinformatic File Formats

	* base number convention
	* BED
	* interval-list
	* GTF/GFF
	
	* BEDTOOLS

* RNAseq quantitation

	* HTseq

* Peak detection

	* MACS

## Lab 3:

### (Prog) Code a simple peak detector

### (Bioinfo) Processes SAM files with Samtools/PICARD	

### (Bioinfo) HTSeq

### (R) Rsamtools


# Module 4: Browsers/Annotation

## Lecture 4:

### UCSC Browser

* Introduction

* BLAT

* Table Browser

* Liftover

### IGV

## Lab 4:

### (Bioinfo) Load/visualize annotation in UCSC 

### (Bioinfo) Load/visualize annotations in IGV

### (All) Open Q&A
